src/main/java/com/deavensoft/paketomat/dispatcher/DispatcherServiceImpl.java

```diff
@@ -1,27 +1,142 @@
         this.dispatcherRepository = dispatcherRepository;
+        this.userService = userService;
+        this.emailService = emailService;
     }
-    public List<DispatcherModel> findAllDispatchers(){
+
+    public List<Dispatcher> findAllDispatchers() {
         return dispatcherRepository.findAll();
     }
-    public void saveDispatcher(DispatcherModel newDispatcher){
+
+    public void saveDispatcher(Dispatcher newDispatcher) {
         dispatcherRepository.save(newDispatcher);
     }
-    public Optional<DispatcherModel> findDispatcherById(Long id){
-       return  dispatcherRepository.findById(id);
+
+    public Optional<Dispatcher> findDispatcherById(Long id) {
+        return dispatcherRepository.findById(id);
     }
-    public void deleteDispatcherById(Long id){
+
+    public void deleteDispatcherById(Long id) {
         dispatcherRepository.deleteById(id);
     }
+
+    @Override
+    public void delieverPackage(Package newPackage) throws IOException, NoSuchUserException {
+        Optional<User> user = userService.findUserById(newPackage.getUser().getId());
+        if(user.isPresent()){
+            String[] parts = user.get().getAddress().split(",", 2);
+            String cityUser = parts[1];
+            findNearestCity(cityUser, newPackage);
+            sendMailToCourier();
+        }else
+            throw new NoSuchUserException("There is no user with id " + newPackage.getUser().getId(), HttpStatus.NOT_FOUND, 404);
+    }
+
+    public void findNearestCity(String address, Package newPackage) throws IOException {
+        City city = null;
+        double minDistance = 1000.0;
+        double distance;
+        ArrayList<City> cities = new ArrayList<>(Center.cities);
+        while (true) {
+            for (City c : cities) {
+                if (c.getPopulation() > MIN_POPULATION) {
+                    distance = findDistance(c.getName(), address);
+                    if (distance < minDistance) {
+                        minDistance = distance;
+                        city = c;
+                    }
+                }
+            }
+            if (city!= null && checkIfAvaiable(newPackage, city))
+                break;
+            cities.remove(city);
+            minDistance = 1000.0;
+        }
+
+    }
+
+    public double findDistance(String cityPaketomat, String cityReciever) throws IOException {
+        OkHttpClient client = new OkHttpClient();
+
+        String newUrl = url.replace("{cityReciever}", cityReciever);
+        String newURL = newUrl.replace("{cityPaketomat}", cityPaketomat);
+        Request request = new Request.Builder()
+                .url(newURL)
+                .get()
+                .addHeader(keyName, keyValue)
+                .addHeader(hostName, hostValue)
+                .build();
+
+        Response response = client.newCall(request).execute();
+        return calculateDistance(response.body().string());
+    }
+
+    public double calculateDistance(String text) {
```

Comment:
What will happen if a response `response.body().string()` doesn't return the expected results?

Reply:
Added exception